### ISSUE TITLE
Unable to restore `v1beta1/Deployment` on kubernetes 1.16+

### ISSUE THREAD

**Issue by Danil-Grigorev** [vmware-tanzu/velero#2310]
**What steps did you take and what happened:**

Velero does not allow to upgrade resource api version on restore, which leads to errors attempting to restore something like `v1beta1/Deployment` on a kubernetes 1.16+ cluster. This leads to a restore going into `PartitiallyFailed` state, an a following error message in the logs:

```
time="2020-03-02T12:02:17Z" level=info msg="error restoring nginx-deployment: the server could not find the requested resource" logSource="pkg/restore/restore.go:1079" restore=openshift-migration/28247050-5c05-11ea-b46b-5b815f8210b4-9rrtc
```

**What did you expect to happen:**

A successful restore with in-place upgrade (something like `kubectl convert`), or a soft blocker on restore, saying this is not possible (dry-run approach? https://github.com/vmware-tanzu/velero/issues/448)

**The output of the following commands will help us better understand what's going on**:

https://gist.github.com/Danil-Grigorev/2e0c3c7e7b13e72c6acf804ff6effb59

**Anything else you would like to add:**

Related discussion https://github.com/konveyor/mig-controller/issues/432

**Environment:**

- Velero version (use `velero version`):  1.2
- Kubernetes version (use `kubectl version`): v1.7 to v1.17.1

**Comment by sseago** [comment]
This is an issue for several resource types: https://kubernetes.io/blog/2019/07/18/api-deprecations-in-1-16/

**Comment by nrb** [comment]
This is correct, Velero does not automatically upgrade API versions, as it can't know all changes without someone writing plugins for each type. The API version must exist on both the backup and restore cluster, as documented at https://velero.io/docs/v1.2.0/how-velero-works/#backed-up-api-versions

I see that you performed a back up on v1.7 and restored on v1.17. `apps/v1` was not available on 1.7, and `v1beta1` is not available post 1.16.

We have some ideas for getting plugins written for larger Kubernetes version jumps, but they're in the early stages now. This basically involves having groups of `RestoreItemAction` plugins that take API types and provide the necessary translations from the older, deprecated/removed versions to the new versions.

`kubectl convert` works right now because it has access to a lot of the Kubernetes internal code, but it won't be around forever - see [this warning](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/convert/convert.go#L146) in their code. This is also code Velero likely can't import.

So to summarize, unfortunately this is known and expected behavior in Velero right now.  However, I would agree that it's a reasonable request to ask that we support backed up resources for a wider window of Kubernetes versions.

**Comment by brito-rafa** [comment]
Please see PR for the *backup* of multiple versions: https://github.com/vmware-tanzu/velero/pull/2373

I wrote a quick script to test the prototype, feel free to use it:
https://github.com/brito-rafa/velero-test/blob/master/2251-dev-test.sh

Please note the restore part is under discussion. Recommend you join the Velero community call to participate.

**Comment by nrb** [comment]
Closing this in favor of #2251 and #2551
